src/request_executor.cc

```diff
@@ -211,25 +272,36 @@ ResponseAlloc(
       } break;
 #ifdef TRITON_ENABLE_GPU
       case TRITONSERVER_MEMORY_GPU: {
-        auto err = cudaSetDevice(*actual_memory_type_id);
-        if ((err != cudaSuccess) && (err != cudaErrorNoDevice) &&
-            (err != cudaErrorInsufficientDriver)) {
-          return TRITONSERVER_ErrorNew(
-              TRITONSERVER_ERROR_INTERNAL,
-              std::string(
-                  "unable to set current CUDA device: " +
-                  std::string(cudaGetErrorString(err)))
-                  .c_str());
-        }
+        if (!IsAllocateGPUBufferSuccess(
+                *actual_memory_type_id, buffer, byte_size)) {
+          int num_devices = 0;
+          cudaError_t err = cudaGetDeviceCount(&num_devices);
+          if (err != cudaSuccess) {
+            return TRITONSERVER_ErrorNew(
+                TRITONSERVER_ERROR_INTERNAL,
+                std::string(
+                    "cudaGetDeviceCount failed: " +
+                    std::string(cudaGetErrorString(err)))
+                    .c_str());
+          }
 
-        err = cudaMalloc(buffer, byte_size);
-        if (err != cudaSuccess) {
-          return TRITONSERVER_ErrorNew(
-              TRITONSERVER_ERROR_INTERNAL,
-              std::string(
-                  "cudaMalloc failed: " + std::string(cudaGetErrorString(err)))
-                  .c_str());
+          bool is_buffer_allocated = false;
+          for (int i = 0; i < num_devices; i++) {
+            if (i != *actual_memory_type_id) {
+              if (IsAllocateGPUBufferSuccess(i, buffer, byte_size)) {
+                is_buffer_allocated = true;
+                break;
+              }
+            }
+          }
+
+          if (!is_buffer_allocated) {
+            return TRITONSERVER_ErrorNew(
+                TRITONSERVER_ERROR_INTERNAL,
+                "unable to allocate buffer on any available CUDA device");
+          }
```

Comment:
@Tabrizian agreed. Allocating additional memory on the other GPUs could inadvertently trigger memory failure on the other running BLS models that could have continued working just fine otherwise. CPU fallback is much better, but I'm not sure if it's ideal. It's better to give the user the option to choose between CPU fallback and upfront model failure. Sometimes you need to do a thorough benchmark on the entire configuration, and you'd rather know at the benchmark stage whether some Triton configuration is stable and working. Falling back to CPU as default could make it harder to catch those out-of-memory situations early on. But it's not really that difficult to achieve under the current design. The developer could simply verify every returned tensor is on the same GPU device and throw an exception when it's not. The code gets a bit dirtier when you have several internal calls happening, but it's doable.

Reply:
Talked to Iman offline, and we decided to simply return an error here as it may be more intuitive.

> It's better to give the user the option to choose between CPU fallback and upfront model failure.

Great suggestion. This can be an enhancement in the future.